hearth/lib/hearth/http/client.rb

```diff
@@ -162,3 +180,39 @@ def http_proxy_parts
     end
   end
 end
+
+# These patches are based on resolv-replace
+# https://github.com/ruby/ruby/blob/master/lib/resolv-replace.rb
+# We cannot require resolv-replace because it would change DNS resolution
+# globally. When opening an HTTP request, we will set a thread local variable
+# to enable custom DNS resolution, and then disable it after the request is
+# complete. When the thread local variable is not set, we will use the default
+# Ruby DNS resolution, which may be Resolv or the system resolver.
+
+# Patch IPSocket
+class << IPSocket
+  alias original_hearth_getaddress getaddress
+
+  def getaddress(host)
```

Comment:
Looking at this code: https://github.com/ruby/ruby/blob/55c771c302f94f1d1d95bf41b42459b4d2d1c337/ext/socket/ipsocket.c#L97

It looks like socket's dns resolution will loop through all of the returned addresses and return the first "healthy" one (where we can create a successful socket connection) - I think.  With this method we're loosing out on that.  Is that a concern?

Reply:
Yeah, we resolve the address and then it's passed into that method. The IP address will return itself, giving a single item to loop through and use. I guess it's a concern, but the requirement is that we need DNS control at the application layer. Unless you see a better option?

Reply:
No - I don't really see any better options - I think its reasonable for what we need to do.